PR: https://github.com/odoo/odoo/pull/218377

From: fc66e2d4eb638f1486e69cd5920f02c787055da1
From: Pieter Claeys (clpi)
Date: 2025-09-06 12:43:45

Breaking data model changes scores: del:1 + add:0, change matches:
-    display_action_record_components = fields.Selection(

Total Changes: 580

[IMP] mrp_subcontracting: new subcontracting

We introduce a new logic and UX for subcontracting, now:
* Subcontracted receipts have a detailed operations popup like any other
  move
* A subcontract receipt for an untracked product has one SBC MO
* A subcontract receipt for a tracked product has one SBC MO per
  lot/serial on its move lines
* Clicking Register Components on a subcontract receipt move will show a
  simplified version of the MO form view (in case of 1 linked SBC MO) or
  list view (in case of multiple linked SBC MOs)
* The split mechanism for MOs is used to enforce sync between the move
  lines of the subcontract receipt and the linked SBC MOs
* We enforce lot M2O fields for subcontracting receipts (over lot name
  text fields)

task-4321311

closes odoo/odoo#218377

Related: odoo/enterprise#89965
Related: odoo/upgrade#8041
Signed-off-by: William Henrotin (whe) <whe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp_subcontracting/models/mrp_production.py
+++ b/addons/mrp_subcontracting/models/mrp_production.py
@@ -63,45 +63,38 @@ class MrpProduction(models.Model):
                 res &= super(MrpProduction, production).write({**vals, 'date_start': date_start_map[production]})
             return res
 
-        return super().write(vals)
+        old_lots = [mo.lot_producing_ids for mo in self]
+        if self.env.context.get('mrp_subcontracting') and 'product_qty' in vals:
+            for mo in self:
+                self.sudo().env['change.production.qty'].with_context(skip_activity=True, mrp_subcontracting=False).create([{
+                    'mo_id': mo.id,
+                    'product_qty': vals['product_qty'],
+                }]).change_prod_qty()
+                mo.action_assign()
+
+        res = super().write(vals)
+
+        if self.env.context.get('mrp_subcontracting') and ('product_qty' in vals or 'lot_producing_ids' in vals):
+            for mo, old_lot in zip(self, old_lots):
+                sbc_move = mo._get_subcontract_move()
+                if not sbc_move:
+                    continue
+                if mo.product_tracking in ('lot', 'serial'):
+                    sbc_move_lines = sbc_move.move_line_ids.filtered(lambda m: m.lot_id == old_lot)
+                    sbc_move_line = sbc_move_lines[0]
+                    sbc_move_line.quantity = mo.product_qty
+                    sbc_move_line.lot_id = mo.lot_producing_ids
+                    sbc_move_lines[1:].unlink()
+                else:
+                    sbc_move.quantity = mo.product_qty
+
+        return res
 
     def action_merge(self):
         if any(production._get_subcontract_move() for production in self):
             raise ValidationError(_("Subcontracted manufacturing orders cannot be merged."))
         return super().action_merge()
 
-    def subcontracting_record_component(self):
-        self.ensure_one()
-        self.move_raw_ids.picked = True
-        if not self._get_subcontract_move():
-            raise UserError(_("This MO isn't related to a subcontracted move"))
-        if self.product_uom_id.is_zero(self.qty_producing):
-            return {'type': 'ir.actions.act_window_close'}
-
-        if self.move_raw_ids and not any(self.move_raw_ids.mapped('quantity')):
-            raise UserError(_("You must indicate a non-zero amount consumed for at least one of your components"))
-        consumption_issues = self._get_consumption_issues()
-        if consumption_issues:
-            return self._action_generate_consumption_wizard(consumption_issues)
-        self.sudo()._update_finished_move()  # Portal user may need sudo rights to update pickings
-        self.subcontracting_has_been_recorded = True
-
-        quantity_issues = self._get_quantity_produced_issues()
-        if quantity_issues:
-            backorder = self.sudo()._split_productions()[1:]
-            # No qty to consume to avoid propagate additional move
-            # TODO avoid : stock move created in backorder with 0 as qty
-            backorder.move_raw_ids.filtered(lambda m: m.additional).product_uom_qty = 0.0
-
-            backorder.qty_producing = backorder.product_qty
-            backorder._set_qty_producing()
-
-            self.product_qty = self.qty_producing
-            action = self._get_subcontract_move().filtered(lambda m: m.state not in ('done', 'cancel'))._action_record_components()
-            action['res_id'] = backorder.id
-            return action
-        return {'type': 'ir.actions.act_window_close'}
-
     def pre_button_mark_done(self):
         if self._get_subcontract_move():
             return super(MrpProduction, self.with_context(skip_consumption=True)).pre_button_mark_done()
@@ -110,74 +103,12 @@ class MrpProduction(models.Model):
     def _should_postpone_date_finished(self, date_finished):
         return super()._should_postpone_date_finished(date_finished) and not self._get_subcontract_move()
 
-    def _update_finished_move(self):
-        """ After producing, set the move line on the subcontract picking. """
-        self.ensure_one()
-        subcontract_move_id = self._get_subcontract_move().filtered(lambda m: m.state not in ('done', 'cancel'))
-        if subcontract_move_id:
-            quantity = self.qty_producing
-            if self.lot_producing_ids:
-                move_lines = subcontract_move_id.move_line_ids.filtered(lambda ml: not ml.picked and ml.lot_id in self.lot_producing_ids or not ml.lot_id)
-            else:
-                move_lines = subcontract_move_id.move_line_ids.filtered(lambda ml: not ml.picked and not ml.lot_id)
-            # Update reservation and quantity done
-            for ml, lot_id in zip(move_lines, self.lot_producing_ids or ([self.env['stock.lot']] * len(move_lines))):
-                rounding = ml.product_uom_id.rounding
-                if float_compare(quantity, 0, precision_rounding=rounding) <= 0:
-                    break
-                quantity_to_process = min(quantity, ml.quantity)
-                quantity -= quantity_to_process
-
-                # on which lot of finished product
-                if float_compare(quantity_to_process, ml.quantity, precision_rounding=rounding) >= 0:
-                    ml.write({
-                        'quantity': quantity_to_process,
-                        'picked': True,
-                        'lot_id': lot_id.id,
-                    })
-                else:
-                    ml.write({
-                        'quantity': quantity_to_process,
-                        'picked': True,
-                        'lot_id': lot_id.id,
-                    })
-
-            if self.product_uom_id.compare(quantity, 0) > 0:
-                self.env['stock.move.line'].create({
-                    'move_id': subcontract_move_id.id,
-                    'picking_id': subcontract_move_id.picking_id.id,
-                    'product_id': self.product_id.id,
-                    'location_id': subcontract_move_id.location_id.id,
-                    'location_dest_id': subcontract_move_id.location_dest_id.id,
-                    'product_uom_id': self.product_uom_id.id,
-                    'quantity': quantity,
-                    'picked': True,
-                    'lot_id': self.lot_producing_ids and self.lot_producing_ids[:1].id,
-                })
-            if not self._get_quantity_to_backorder():
-                subcontract_move_id.move_line_ids.filtered(lambda ml: not ml.picked).unlink()
-                subcontract_move_id._recompute_state()
-
-    def _subcontracting_filter_to_done(self):
-        """ Filter subcontracting production where composant is already recorded and should be consider to be validate """
-        def filter_in(mo):
-            if mo.state in ('done', 'cancel'):
-                return False
-            if not mo.subcontracting_has_been_recorded:
-                return False
-            return True
-
-        return self.filtered(filter_in)
-
     def _has_been_recorded(self):
         self.ensure_one()
         if self.state in ('cancel', 'done'):
             return True
         return self.subcontracting_has_been_recorded
 
-    def _has_tracked_component(self):
-        return any(m.has_tracking != 'none' for m in self.move_raw_ids)
-
     def _has_workorders(self):
         if self.subcontractor_id:
             return False
@@ -190,11 +121,19 @@ class MrpProduction(models.Model):
     def _get_writeable_fields_portal_user(self):
         return ['move_line_raw_ids', 'lot_producing_ids', 'subcontracting_has_been_recorded', 'qty_producing', 'product_qty']
 
-    def _subcontract_sanity_check(self):
-        for production in self:
-            if production.product_tracking != 'none' and not self.lot_producing_ids:
-                raise UserError(_('You must enter a serial number for %s', production.product_id.name))
-            for sml in production.move_raw_ids.move_line_ids:
-                if sml.tracking != 'none' and not sml.lot_id:
-                    raise UserError(_('You must enter a serial number for each line of %s', sml.product_id.display_name))
-        return True
+    def action_split_subcontracting(self):
+        self.ensure_one()
+        if not self.lot_producing_ids:
+            raise UserError(_("Please set a lot/serial for the currently opened subcontracting MO first."))
+        move = self._get_subcontract_move()
+        if not move:
+            return False
+        empty_line = move.move_line_ids.filtered(lambda l: not l.lot_id)
+        if not empty_line:
+            move.move_line_ids.create({
+                'product_id': move.product_id.id,
+                'move_id': move.id,
+                'quantity': 1,
+                'lot_id': False,
+            })
+        return move.action_show_subcontract_details(lot_id=False)

--- a/addons/mrp_subcontracting/models/stock_move.py
+++ b/addons/mrp_subcontracting/models/stock_move.py
@@ -16,93 +16,34 @@ class StockMove(models.Model):
         compute='_compute_show_subcontracting_details_visible'
     )
 
-    def _compute_display_assign_serial(self):
-        super(StockMove, self)._compute_display_assign_serial()
-        for move in self:
-            if not move.is_subcontract:
-                continue
-            productions = move._get_subcontract_production()
-            if not productions or move.has_tracking == 'none':
-                continue
-            if productions._has_tracked_component() or productions[:1].consumption != 'strict':
-                move.display_assign_serial = False
-
     def _compute_show_subcontracting_details_visible(self):
         """ Compute if the action button in order to see moves raw is visible """
         self.show_subcontracting_details_visible = False
         for move in self:
             if not move.is_subcontract:
                 continue
-            if move.product_uom.is_zero(move.quantity):
+            if not move.move_line_ids or move.product_uom.is_zero(move.quantity):
                 continue
             productions = move._get_subcontract_production()
-            if not productions or (productions[:1].consumption == 'strict' and not productions[:1]._has_tracked_component()):
+            if not productions:
                 continue
-            move.show_subcontracting_details_visible = move.picked or any(p.subcontracting_has_been_recorded and p.state != 'done' for p in productions)
+            move.show_subcontracting_details_visible = True
 
-    def _compute_show_details_visible(self):
-        """ If the move is subcontract and the components are tracked. Then the
-        show details button is visible.
-        """
-        res = super(StockMove, self)._compute_show_details_visible()
-        for move in self:
-            if not move.is_subcontract:
-                continue
-            if self.env.user._is_portal():
-                move.show_details_visible = any(not p._has_been_recorded() for p in move._get_subcontract_production())
-                continue
-            productions = move._get_subcontract_production()
-            if not productions._has_tracked_component() and productions[:1].consumption == 'strict':
-                continue
-            move.show_details_visible = True
-        return res
+    @api.depends('is_subcontract')
+    def _compute_show_info(self):
+        super()._compute_show_info()
+        subcontract_moves = self.filtered(lambda m: m.is_subcontract and m.show_lots_text)
+        subcontract_moves.show_lots_text = False
+        subcontract_moves.show_lots_m2o = True
 
+    @api.depends('is_subcontract')
     def _compute_is_quantity_done_editable(self):
-        not_editable = self.filtered(lambda m: m.is_subcontract and (m.move_orig_ids.production_id._has_tracked_component() or m.has_tracking != "none"))
-        not_editable.is_quantity_done_editable = False
-        super(StockMove, self - not_editable)._compute_is_quantity_done_editable()
-
-    def _set_quantity_done(self, qty):
-        to_set_moves = self
+        done_moves = self.env['stock.move']
         for move in self:
-            if move.is_subcontract and move._subcontracting_possible_record():
-                # If 'done' quantity is changed through the move, record components as if done through the wizard.
-                move._auto_record_components(qty)
-                to_set_moves -= move
-        if to_set_moves:
-            super(StockMove, to_set_moves)._set_quantity_done(qty)
-
-    def _set_quantity(self):
-        to_set_moves = self
-        for move in self:
-            if move.is_subcontract and move._subcontracting_possible_record():
-                move_line_quantities = sum(move.move_line_ids.filtered(lambda ml: ml.picked).mapped('quantity'))
-                delta_qty = move.quantity - move_line_quantities
-                if move.product_uom.compare(delta_qty, 0) > 0:
-                    move._auto_record_components(delta_qty)
-                    to_set_moves -= move
-                elif move.product_uom.compare(delta_qty, 0) < 0:
-                    move.with_context(transfer_qty=True)._reduce_subcontract_order_qty(abs(delta_qty))
-        if to_set_moves:
-            super(StockMove, to_set_moves)._set_quantity()
-
-    def _auto_record_components(self, qty):
-        self.ensure_one()
-        subcontracted_productions = self._get_subcontract_production()
-        production = subcontracted_productions.filtered(lambda p: not p._has_been_recorded())[-1:]
-        if not production:
-            # If new quantity is over the already recorded quantity and we have no open production, then create a new one for the missing quantity.
-            production = subcontracted_productions[-1:]
-            production = production.sudo().with_context(allow_more=True)._split_productions({production: [production.qty_producing, qty]})[-1:]
-        qty = self.product_uom._compute_quantity(qty, production.product_uom_id)
-        production.qty_producing = qty
-        if production.product_uom_id.compare(production.qty_producing, production.product_qty) > 0:
-            self.env['change.production.qty'].with_context(skip_activity=True).create({
-                'mo_id': production.id,
-                'product_qty': qty
-            }).change_prod_qty()
-        production._set_qty_producing()
-        production.with_context(cancel_backorder=False).subcontracting_record_component()
+            if move.is_subcontract:
+                move.is_quantity_done_editable = False
+                done_moves |= move
+        return super(StockMove, self - done_moves)._compute_is_quantity_done_editable()
 
     def copy_data(self, default=None):
         default = dict(default or {})
@@ -118,11 +59,6 @@ class StockMove(models.Model):
         subcontract order to the new quantity.
         """
         self._check_access_if_subcontractor(vals)
-        if 'product_uom_qty' in vals and self.env.context.get('cancel_backorder') is not False and not self.env.context.get('extra_move_mode'):
-            self.filtered(
-                lambda m: m.is_subcontract and m.state not in ['draft', 'cancel', 'done']
-                and m.product_uom.compare(m.product_uom_qty, vals['product_uom_qty']) != 0
-            )._update_subcontract_order_qty(vals['product_uom_qty'])
         res = super().write(vals)
         if 'date' in vals:
             for move in self:
@@ -145,37 +81,48 @@ class StockMove(models.Model):
         subcontracted product. Otherwise use standard behavior.
         """
         self.ensure_one()
-        if self.state != 'done' and (self._subcontrating_should_be_record() or self._subcontrating_can_be_record()):
-            return self._action_record_components()
-        action = super(StockMove, self).action_show_details()
-        if self.is_subcontract and all(p._has_been_recorded() for p in self._get_subcontract_production()):
-            action['views'] = [(self.env.ref('stock.view_stock_move_operations').id, 'form')]
-            action['context'].update({
-                'show_lots_m2o': self.has_tracking != 'none',
-                'show_lots_text': False,
-            })
-        elif self.env.user._is_portal():
-            action['views'] = [(self.env.ref('mrp_subcontracting.mrp_subcontracting_view_stock_move_operations').id, 'form')]
-        return action
+        if self.is_subcontract:
+            action = super(StockMove, self.with_context(force_lot_m2o=True)).action_show_details()
+            if self.env.user._is_portal():
+                action['views'] = [(self.env.ref('mrp_subcontracting.mrp_subcontracting_view_stock_move_operations').id, 'form')]
+            return action
+        return super().action_show_details()
 
-    def action_show_subcontract_details(self):
+    def action_show_subcontract_details(self, lot_id=None):
         """ Display moves raw for subcontracted product self. """
-        moves = self._get_subcontract_production().move_raw_ids.filtered(lambda m: m.state != 'cancel')
-        list_view = self.env.ref('mrp_subcontracting.mrp_subcontracting_move_tree_view')
-        form_view = self.env.ref('mrp_subcontracting.mrp_subcontracting_move_form_view')
-        ctx = dict(self.env.context, search_default_by_product=True)
+        productions = self._get_subcontract_production().filtered(lambda m: m.state != 'cancel')
+        if lot_id is not None:
+            if lot_id:
+                productions = productions.filtered(lambda p: p.lot_producing_ids and p.lot_producing_ids[0] == self.env['stock.lot'].browse(lot_id))
+            else:
+                productions = productions.filtered(lambda p: not p.lot_producing_ids)
+        ctx = {"mrp_subcontracting": True}
         if self.env.user._is_portal():
-            form_view = self.env.ref('mrp_subcontracting.mrp_subcontracting_portal_move_form_view')
+            form_view_id = self.env.ref('mrp_subcontracting.mrp_production_subcontracting_portal_form_view')
             ctx.update(no_breadcrumbs=False)
-        return {
-            'name': _('Raw Materials for %s', self.product_id.display_name),
+        else:
+            form_view_id = self.env.ref('mrp_subcontracting.mrp_production_subcontracting_form_view')
+        action = {
             'type': 'ir.actions.act_window',
-            'res_model': 'stock.move',
-            'views': [(list_view.id, 'list'), (form_view.id, 'form')],
+            'res_model': 'mrp.production',
             'target': 'current',
-            'domain': [('id', 'in', moves.ids)],
             'context': ctx
         }
+        if len(productions) > 1:
+            action.update({
+                'name': _('Subcontracting MOs'),
+                'views': [
+                    (self.env.ref('mrp_subcontracting.mrp_production_subcontracting_tree_view').id, 'list'),
+                    (form_view_id.id, 'form'),
+                ],
+                'domain': [('id', 'in', productions.ids)],
+            })
+        else:
+            action.update({
+                'views': [(form_view_id.id, 'form')],
+                'res_id': productions.id,
+            })
+        return action
 
     def _action_cancel(self):
         productions_to_cancel_ids = OrderedSet()
@@ -223,26 +170,6 @@ class StockMove(models.Model):
             self.env['stock.picking'].concat(*list(subcontract_details_per_picking.keys())).action_assign()
         return res
 
-    def _action_record_components(self):
-        self.ensure_one()
-        production = self._get_subcontract_production()[-1:]
-        view = self.env.ref('mrp_subcontracting.mrp_production_subcontracting_form_view')
-        if self.env.user._is_portal():
-            view = self.env.ref('mrp_subcontracting.mrp_production_subcontracting_portal_form_view')
-        context = dict(self.env.context)
-        context.pop('skip_consumption', False)
-        return {
-            'name': _('Subcontract'),
-            'type': 'ir.actions.act_window',
-            'view_mode': 'form',
-            'res_model': 'mrp.production',
-            'views': [(view.id, 'form')],
-            'view_id': view.id,
-            'target': 'new',
-            'res_id': production.id,
-            'context': context,
-        }
-
     def _get_subcontract_bom(self):
         self.ensure_one()
         bom = self.env['mrp.bom'].sudo()._bom_subcontract_find(
@@ -254,22 +181,9 @@ class StockMove(models.Model):
         )
         return bom
 
-    def _subcontrating_should_be_record(self):
-        return self._get_subcontract_production().filtered(lambda p: not p._has_been_recorded() and p._has_tracked_component())
-
-    def _subcontrating_can_be_record(self):
-        return self._get_subcontract_production().filtered(lambda p: not p._has_been_recorded() and p.consumption != 'strict')
-
-    def _subcontracting_possible_record(self):
-        return self._get_subcontract_production().filtered(lambda p: p._has_tracked_component() or p.consumption != 'strict')
-
     def _get_subcontract_production(self):
         return self.filtered(lambda m: m.is_subcontract).move_orig_ids.production_id
 
-    # TODO: To be deleted, use self._get_subcontract_production()._has_tracked_component() instead
-    def _has_tracked_subcontract_components(self):
-        return any(m.has_tracking != 'none' for m in self._get_subcontract_production().move_raw_ids)
-
     def _prepare_move_split_vals(self, qty):
         vals = super(StockMove, self)._prepare_move_split_vals(qty)
         vals['location_id'] = self.location_id.id
@@ -291,39 +205,6 @@ class StockMove(models.Model):
     def _get_available_move_lines(self, assigned_moves_ids, partially_available_moves_ids):
         return super(StockMove, self.filtered(lambda m: not m.is_subcontract))._get_available_move_lines(assigned_moves_ids, partially_available_moves_ids)
 
-    def _update_subcontract_order_qty(self, new_quantity):
-        for move in self:
-            quantity_to_remove = move.product_uom_qty - new_quantity
-            if not move.product_uom.is_zero(quantity_to_remove):
-                move._reduce_subcontract_order_qty(quantity_to_remove)
-
-    def _reduce_subcontract_order_qty(self, quantity_to_remove):
-        self.ensure_one()
-        productions = self.move_orig_ids.production_id.filtered(lambda p: p.state not in ('done', 'cancel'))[::-1]
-        wip_production = productions[0] if self.env.context.get('transfer_qty') and len(productions) > 1 else self.env['mrp.production']
-
-        # Transfer removed qty to WIP production
-        if wip_production:
-            self.env['change.production.qty'].with_context(skip_activity=True).create({
-                'mo_id': wip_production.id,
-                'product_qty': wip_production.product_qty + quantity_to_remove
-            }).change_prod_qty()
-
-        # Cancel productions until reach new_quantity
-        for production in (productions - wip_production):
-            if float_compare(quantity_to_remove, production.product_qty, precision_rounding=production.product_uom_id.rounding) >= 0:
-                quantity_to_remove -= production.product_qty
-                production.with_context(skip_activity=True).action_cancel()
-            else:
-                if production.product_uom_id.is_zero(quantity_to_remove):
-                    # No need to do change_prod_qty for no change at all.
-                    break
-                self.env['change.production.qty'].with_context(skip_activity=True).create({
-                    'mo_id': production.id,
-                    'product_qty': production.product_qty - quantity_to_remove
-                }).change_prod_qty()
-                break
-
     def _check_access_if_subcontractor(self, vals):
         if self.env.user._is_portal() and not self.env.su:
             if vals.get('state') == 'done':
@@ -337,3 +218,84 @@ class StockMove(models.Model):
                 and self.origin_returned_move_id.is_subcontract
                 and self.location_dest_id.id == subcontracting_location.id
         )
+
+    def _sync_subcontracting_productions(self):
+        """
+            Enforce the relationship between subcontracting receipt moves and their respective subcontracting productions.
+            * For untracked moves:
+                * There will always be only 1 production.
+                * Updating the move quantity will update the production quantity.
+            * For tracked moves:
+                * There will be 1 production for every lot on this move.
+                * This method will enforce the synchronisation between the total quantity per lot on the move and the linked productions.
+                * The split mechanism for productions will be used to create new subcontracting MOs.
+                * We take care to always keep at least 1 subcontracting production linked to the subcontracting receipt.
+                  This ensures there will always be a production available for splitting.
+        """
+        for move in self:
+            if move.has_tracking == 'none':
+                mo = move._get_subcontract_production()
+                if mo and mo.product_uom_id.compare(mo.product_qty, move.quantity) != 0:
+                    self.sudo().env['change.production.qty'].with_context(skip_activity=True).create([{
+                        'mo_id': mo.id,
+                        'product_qty': move.quantity or move.product_uom_qty,
+                    }]).change_prod_qty()
+                    mo.action_assign()
+            else:
+                productions = move._get_subcontract_production()
+                if not productions:
+                    continue
+                qty_by_lot = dict(move.move_line_ids._read_group([('move_id', '=', move.id)], ['lot_id'], ['quantity:sum']))
+                open_mo = productions.filtered(lambda p: not p.lot_producing_ids and not p._has_been_recorded())
+                if open_mo and qty_by_lot:
+                    # No MOs created yet for the lots, split them from the default 'open' MO
+                    productions = open_mo.sudo().with_context(allow_more=True)._split_productions({open_mo: list(qty_by_lot.values())}, cancel_remaining_qty=True, skip_procurement=False)
+                    for production, lot_id in zip(productions, qty_by_lot.keys()):
+                        production.lot_producing_ids = lot_id
+                        production.subcontracting_has_been_recorded = True
+                else:
+                    mos_to_assign = self.env['mrp.production']
+                    # MOs already created for lots, update to enforce sync:
+                    # 1. Ensure quantities of linked MOs still match the quantities on the move
+                    mos_to_create = {}  # lot -> qty
+                    for lot_id, ml_qty in qty_by_lot.items():
+                        lot_mo = productions.filtered(lambda p: (p.lot_producing_ids and p.lot_producing_ids[0] == lot_id) or (not lot_id and not p.lot_producing_ids))
+                        if not lot_mo:
+                            mos_to_create[lot_id] = ml_qty
+                        elif lot_mo.product_uom_id.compare(lot_mo.product_qty, ml_qty) != 0:
+                            self.sudo().env['change.production.qty'].with_context(skip_activity=True).create([{
+                                'mo_id': lot_mo.id,
+                                'product_qty': ml_qty
+                            }]).change_prod_qty()
+                            mos_to_assign |= lot_mo
+
+                    # 2. Create new MOs where needed, by splitting them from an existing subcontracting MO
+                    if mos_to_create:
+                        production_to_split = move._get_subcontract_production()[0]
+                        new_mos = production_to_split.sudo().with_context(allow_more=True, mrp_subcontracting=False)._split_productions({
+                            production_to_split: [production_to_split.product_qty] + list(mos_to_create.values())
+                        }, cancel_remaining_qty=True)[1:]
+                        mos_to_assign |= new_mos
+                        for mo, lot_id in zip(new_mos, mos_to_create.keys()):
+                            mo.lot_producing_ids = lot_id
+                            mo.subcontracting_has_been_recorded = True
+
+                    # 3. Delete 'orphan' MOs with lot not linked to any move line
+                    productions = move._get_subcontract_production()
+                    orphan_productions = productions.filtered(lambda p: (p.lot_producing_ids and p.lot_producing_ids[0] not in qty_by_lot) or (not p.lot_producing_ids and self.env['stock.lot'] not in qty_by_lot))
+                    if len(productions) == len(orphan_productions):
+                        # Make sure not to delete all MOs, leave 1 subcontracting MO as 'open' MO for splitting later
+                        production_to_keep = orphan_productions[-1]
+                        production_to_keep.subcontracting_has_been_recorded = False
+                        production_to_keep.lot_producing_ids = False
+                        orphan_productions = orphan_productions[:-1]
+                    if orphan_productions:
+                        orphan_productions.with_context(skip_activity=True).unlink()
+                        productions -= orphan_productions
+
+                    mos_to_assign.action_assign()
+
+    def _generate_serial_numbers(self, next_serial, next_serial_count=False, location_id=False):
+        if self.is_subcontract:
+            return super(StockMove, self.with_context(force_lot_m2o=True))._generate_serial_numbers(next_serial, next_serial_count, location_id)
+        return super()._generate_serial_numbers(next_serial, next_serial_count, location_id)

--- a/addons/mrp_subcontracting/models/stock_move_line.py
+++ b/addons/mrp_subcontracting/models/stock_move_line.py
@@ -20,10 +20,19 @@ class StockMoveLine(models.Model):
         return res
 
     def write(self, vals):
-        for move_line in self:
-            if vals.get('lot_id') and move_line.move_id.is_subcontract and move_line.location_id.is_subcontracting_location:
-                # Update related subcontracted production to keep consistency between production and reception.
-                subcontracted_production = move_line.move_id._get_subcontract_production().filtered(lambda p: p.state not in ('done', 'cancel') and move_line.lot_id.id in p.lot_producing_ids.ids)
-                if subcontracted_production:
-                    subcontracted_production.lot_producing_ids = [vals['lot_id']]
-        return super().write(vals)
+        res = super().write(vals)
+        if not self.env.context.get('mrp_subcontracting') and ('quantity' in vals or 'lot_id' in vals):
+            self.move_id.filtered(lambda m: m.is_subcontract)._sync_subcontracting_productions()
+        return res
+
+    def unlink(self):
+        moves_to_sync = self.move_id.filtered(lambda m: m.is_subcontract)
+        res = super().unlink()
+        moves_to_sync._sync_subcontracting_productions()
+        return res
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        res.move_id.filtered(lambda m: m.is_subcontract)._sync_subcontracting_productions()
+        return res

--- a/addons/mrp_subcontracting/models/stock_picking.py
+++ b/addons/mrp_subcontracting/models/stock_picking.py
@@ -12,24 +12,6 @@ from dateutil.relativedelta import relativedelta
 class StockPicking(models.Model):
     _inherit = 'stock.picking'
 
-    display_action_record_components = fields.Selection(
-        [('hide', 'Hide'), ('facultative', 'Facultative'), ('mandatory', 'Mandatory')],
-        compute='_compute_display_action_record_components')
-
-    @api.depends('state', 'move_ids')
-    def _compute_display_action_record_components(self):
-        self.display_action_record_components = 'hide'
-        for picking in self:
-            # Hide if not encoding state or it is not a subcontracting picking
-            if picking.state in ('draft', 'cancel', 'done') or not picking._is_subcontract():
-                continue
-            subcontracted_moves = picking.move_ids.filtered(lambda m: m.is_subcontract)
-            if subcontracted_moves._subcontrating_should_be_record():
-                picking.display_action_record_components = 'mandatory'
-                continue
-            if subcontracted_moves._subcontrating_can_be_record():
-                picking.display_action_record_components = 'facultative'
-
     @api.depends('picking_type_id', 'partner_id')
     def _compute_location_id(self):
         super()._compute_location_id()
@@ -48,76 +30,25 @@ class StockPicking(models.Model):
     # -------------------------------------------------------------------------
     def _action_done(self):
         res = super(StockPicking, self)._action_done()
-        for move in self.move_ids:
-            if not move.is_subcontract:
-                continue
-            # Auto set qty_producing/lot_producing_ids of MO wasn't recorded
-            # manually (if the flexible + record_component or has tracked component)
-            productions = move._get_subcontract_production()
-            recorded_productions = productions.filtered(lambda p: p._has_been_recorded())
-            recorded_qty = sum(recorded_productions.mapped('qty_producing'))
-            sm_done_qty = sum(productions._get_subcontract_move().filtered(lambda m: m.picked).mapped('quantity'))
-            rounding = self.env['decimal.precision'].precision_get('Product Unit')
-            if float_compare(move.product_uom_qty, move.quantity, precision_digits=rounding) > 0 and self.env.context.get('cancel_backorder'):
-                move._update_subcontract_order_qty(move.quantity)
-            if float_compare(recorded_qty, sm_done_qty, precision_digits=rounding) >= 0:
-                continue
-            production = productions - recorded_productions
-            if not production:
-                continue
-            if len(production) > 1:
-                raise UserError(_("There shouldn't be multiple productions to record for the same subcontracted move."))
-            # Manage additional quantities
-            quantity_done_move = move.product_uom._compute_quantity(move.quantity, production.product_uom_id)
-            if production.product_uom_id.compare(production.product_qty, quantity_done_move) == -1:
-                change_qty = self.env['change.production.qty'].create({
-                    'mo_id': production.id,
-                    'product_qty': quantity_done_move
-                })
-                change_qty.with_context(skip_activity=True).change_prod_qty()
-            # Create backorder MO for each move lines
-            amounts = [move_line.quantity for move_line in move.move_line_ids]
-            len_amounts = len(amounts)
-            productions = production._split_productions({production: amounts}, set_consumed_qty=True)
-            productions.move_finished_ids.move_line_ids.write({'quantity': 0})
-            for production, move_line in zip(productions, move.move_line_ids):
-                if move_line.lot_id:
-                    production.lot_producing_ids = move_line.lot_id.ids
-                production.qty_producing = production.product_qty
-                production._set_qty_producing()
-            productions[:len_amounts].subcontracting_has_been_recorded = True
-
         for picking in self:
-            productions_to_done = picking._get_subcontract_production()._subcontracting_filter_to_done()
-            productions_to_done._subcontract_sanity_check()
-            if not productions_to_done:
-                continue
-            productions_to_done = productions_to_done.sudo()
-            production_ids_backorder = []
-            if not self.env.context.get('cancel_backorder'):
-                production_ids_backorder = productions_to_done.filtered(lambda mo: mo.state == "progress").ids
-            productions_to_done.with_context(mo_ids_to_backorder=production_ids_backorder).button_mark_done()
+            productions_to_done = picking._get_subcontract_production().sudo()
+            productions_to_done.button_mark_done()
             # For concistency, set the date on production move before the date
             # on picking. (Traceability report + Product Moves menu item)
-            minimum_date = min(picking.move_line_ids.mapped('date'))
             production_moves = productions_to_done.move_raw_ids | productions_to_done.move_finished_ids
-            production_moves.write({'date': minimum_date - timedelta(seconds=1)})
-            production_moves.move_line_ids.write({'date': minimum_date - timedelta(seconds=1)})
+            if production_moves:
+                minimum_date = min(picking.move_line_ids.mapped('date'))
+                production_moves.write({'date': minimum_date - timedelta(seconds=1)})
+                production_moves.move_line_ids.write({'date': minimum_date - timedelta(seconds=1)})
 
         return res
 
-    def action_record_components(self):
-        self.ensure_one()
-        move_subcontracted = self.move_ids.filtered(lambda m: m.is_subcontract)
-        for move in move_subcontracted:
-            production = move._subcontrating_should_be_record()
-            if production:
-                return move._action_record_components()
-        for move in move_subcontracted:
-            production = move._subcontrating_can_be_record()
-            if production:
-                return move._action_record_components()
-        raise UserError(_("Nothing to record"))
+    @api.depends('move_ids.is_subcontract', 'move_ids.has_tracking')
+    def _compute_show_lots_text(self):
+        super()._compute_show_lots_text()
+        for picking in self:
+            if any(move.is_subcontract and move.has_tracking != 'none' for move in picking.move_ids):
+                picking.show_lots_text = False
 
     # -------------------------------------------------------------------------
     # Subcontract helpers
@@ -161,6 +92,9 @@ class StockPicking(models.Model):
         return vals
 
     def _get_subcontract_mo_confirmation_ctx(self):
+        if self._is_subcontract() and not self.env.context.get('cancel_backorder', True):
+            # Do not trigger rules on raw moves when creating backorder for a subcontract receipt.
+            return {'no_procurement': True}
         return {}  # To override in mrp_subcontracting_purchase
 
     def _subcontracted_produce(self, subcontract_details):
@@ -168,8 +102,18 @@ class StockPicking(models.Model):
         group_move = defaultdict(list)
         group_by_company = defaultdict(list)
         for move, bom in subcontract_details:
-            # do not create extra production for move that have their quantity updated
             if move.move_orig_ids.production_id:
+                # Magic spicy sauce for the backorder case:
+                # To ensure correct splitting of the component moves of the SBC MO, we will invoke a split of the SBC
+                # MO here directly and then link the backorder MO to the backorder move.
+                # If we would just run _subcontracted_produce as usual for the newly created SBC receipt move, any
+                # reservations of raw component moves of the SBC MO would not be preserved properly (for example when
+                # using resupply subcontractor on order)
+                production_to_split = move.move_orig_ids[0].production_id
+                original_qty = move.move_orig_ids[0].product_qty
+                move.move_orig_ids = False
+                _, new_mo = production_to_split.with_context(allow_more=True)._split_productions({production_to_split: [original_qty, move.product_qty]})
+                new_mo.move_finished_ids.move_dest_ids = move
                 continue
             quantity = move.product_qty or move.quantity
             if move.product_uom.compare(quantity, 0) <= 0:
@@ -188,7 +132,11 @@ class StockPicking(models.Model):
             all_mo.update(grouped_mo.ids)
 
         all_mo = self.env['mrp.production'].browse(sorted(all_mo))
-        all_mo.with_context(self._get_subcontract_mo_confirmation_ctx()).action_confirm()
+        ctx = self._get_subcontract_mo_confirmation_ctx()
+        all_mo.with_context(ctx).action_confirm()
+        if ctx.get('no_procurement'):
+            # Make sure to check availability to the backorder
+            all_mo.action_assign()
 
         for mo in all_mo:
             move = group_move[mo.procurement_group_id.id][0]
